### PR TITLE
Add dart command line tool for calculating engine hash

### PR DESCRIPTION
**Pull request by jtmcdole**
Useful in tooling, releases, etc - where Dart is available

Example usages:

```bash
# Monorepo world
❯ dart ./dev/tools/bin/engine_hash.dart -s head
226e13826c7253c968d798666f323b1f207979f8

# Non-monorepo world
❯ dart ./dev/tools/bin/engine_hash.dart -s head
Error calculating engine hash: Not in a monorepo
```

### CODE REVIEW

**Comment by jtmcdole** [comment]
@christopherfujino - if you have the dart sdk downloaded already, this will provide you with the signature hash.

**Comment by jtmcdole** [comment]
I'll port the test over from the closed engine pr

File: dev/tools/bin/engine_hash.dart [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import 'package:crypto/crypto.dart';
+
+enum GitRevisionStrategy {
+  mergeBase,
+  head,
+}
+
+final _hashRegex = RegExp(r'^([a-fA-F0-9]+)');
+
+final parser = ArgParser()
```
Comment (bkonyi) [inline dev/tools/bin/engine_hash.dart:21]: How picky are we with conforming to the framework style guide for tooling like this? I personally prefer omitting types on the LHS when the type is obvious, but I know that's not typically what we do for framework / tool code.

File: dev/tools/bin/engine_hash.dart [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    allowedHelp: {
+      'head': 'hash from git HEAD',
+      'mergeBase': 'hash from the merge-base of HEAD and upstream/master',
+    },
+  )
+  ..addFlag('help', abbr: 'h', negatable: false);
+
+Never printHelp({String? error}) {
+  if (error != null) {
+    stdout.writeln(error);
```
Comment (bkonyi) [inline dev/tools/bin/engine_hash.dart:36]: Nit: Should this be on `stderr`?

File: dev/tools/bin/engine_hash.dart [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        return (
+          result: '',
+          error: '''
+Unable to find merge-base hash of the repository:
+${processResult.stderr}''',
+        );
+      }
+
+      final baseHash = _hashRegex.matchAsPrefix(processResult.stdout as String);
+      if (baseHash == null || baseHash.groupCount != 1) {
```
Comment (bkonyi) [inline dev/tools/bin/engine_hash.dart:109]: Nit: this could also be `baseHash?.groupCount != 1`

File: dev/tools/bin/engine_hash.dart [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    return 1;
+  }
+
+  stdout.writeln(result.result);
+
+  return 0;
+}
+
+/// Returns the hash signature for the engine source code.
+Future<({String result, String? error})> engineHash(
```
Comment (bkonyi) [inline dev/tools/bin/engine_hash.dart:80]: Any reason we aren't just throwing an exception and returning a `Future<String>`?

File: dev/tools/bin/engine_hash.dart (hunk 2 above)
Comment (jtmcdole) [reply dev/tools/bin/engine_hash.dart:36]: Yes it should!

**Review by bkonyi — APPROVED** [review]
LGTM!

**Pull request merged**